### PR TITLE
[core] add grpc opencensus plugin

### PR DESCRIPTION
**Pull request by rynewang**
gRPC has a built in plugin to export metrics via opencensus, as detailed in https://github.com/census-instrumentation/opencensus-specs/blob/master/stats/gRPC.md . We already use opencensus to collect Ray metrics, and here we add the gRPC-sourced metrics to the list.

In consideration of potential performance impacts, we disable it by default and can be enabled by component via config `enable_grpc_metrics_collection_for`. Now we only support `gcs`, in following patches we plan to support `raylet` and `core_worker`. The reason we don't support it right away is, we need to configure before any gRPC traffic, but in raylet and in core worker we init our stats *after* a grpc call to gcs/raylet to get the configs updated.

### CODE REVIEW

**Comment by jjyao** [comment]
Nice, it would be very valuable.

**Comment by rynewang** [comment]
it works now!

image

File: dashboard/modules/reporter/reporter_agent.py
Comment (jjyao) [inline dashboard/modules/reporter/reporter_agent.py:378]: Is this needed?

File: src/ray/stats/metric.cc
Comment (jjyao) [inline src/ray/stats/metric.cc:47]: Will we have ip or node id label so that we know which node emits these grpc metrics?

File: dashboard/modules/reporter/reporter_agent.py
Comment (rynewang) [reply dashboard/modules/reporter/reporter_agent.py]: removed.

File: src/ray/stats/metric.cc
Comment (rynewang) [reply src/ray/stats/metric.cc:47]: added the global tags in exporter.

**Comment by rynewang** [comment]
wait, it compiles in my laptop but not in CI

**Review by jjyao — COMMENTED** [review]
Thanks! LGTM but I'll defer to @rkooo567 since he is more familiar with this code.

**Comment by rkooo567** [comment]
can we make the grpc stats off by default (add a system config to enable it)?

Also can you give me a list of grpc stats? I wonder if we should filter some of them since it will be a lot. 

Lastly, can you pick which stats we should track from the multi cloud setup?

**Comment by rynewang** [comment]
> can we make the grpc stats off by default (add a system config to enable it)?

config added.

> Also can you give me a list of grpc stats? I wonder if we should filter some of them since it will be a lot.

https://github.com/census-instrumentation/opencensus-specs/blob/master/stats/gRPC.md

> Lastly, can you pick which stats we should track from the multi cloud setup?

`grpc.io/client/roundtrip_latency` and `grpc.io/server/server_latency` is my special interest

**Comment by rkooo567** [comment]
SGTM. I guess we can also track the completed rpcs with status to understand the network failures

```
Why no error counts?
Error counts can be computed on your metrics backend by totalling the different per-status values.
```

File: python/ray/_private/metrics_agent.py
Comment (rkooo567) [inline python/ray/_private/metrics_agent.py:78]: can we add an unit test for this method? (with an expected output coming from cpp components?)

File: python/ray/_private/metrics_agent.py
Comment (rkooo567) [inline python/ray/_private/metrics_agent.py:88]:
```suggestion
    be > 0[4]. However, gRPC OpenCensus metrics have their first bucket bound == 0 [1].
```

What does [4] mean here?

File: python/ray/_private/metrics_agent.py
Comment (rkooo567) [inline python/ray/_private/metrics_agent.py:215]: Is this requiring high CPU usage? Should we have a simple check like statswith("grpc_") to avoid iterating every metrics?

File: src/ray/stats/metric_exporter.cc
Comment (rkooo567) [inline src/ray/stats/metric_exporter.cc:141]: Is it unit testable in metrics_exporter_test.cc?

**Review by rkooo567 — APPROVED** [review]
LGTM! Excited to have grpc metrics :)! Nit comments on adding unit tests

File: python/ray/_private/metrics_agent.py
Comment (rynewang) [reply python/ray/_private/metrics_agent.py:88]: the fourth reference link.

File: python/ray/_private/metrics_agent.py
Comment (rynewang) [reply python/ray/_private/metrics_agent.py]: test added.

File: src/ray/stats/metric_exporter.cc
Comment (rynewang) [reply src/ray/stats/metric_exporter.cc:141]: yes but let's make it a TODO

Comment (rkooo567) [reply src/ray/stats/metric_exporter.cc:141]: Hmm why? When will you do it?

**Review by rkooo567 — CHANGES_REQUESTED** [review]
TODO

- Allow to choose components (gcs, worker, raylet) to export grpc metrics due to cardinality concern. 

I think we can do it by creating a separate method & call it in the beg of the entrypoint script

**Comment by rynewang** [comment]
Blocked on https://github.com/ray-project/ray/pull/39210 because we need grpc to be upgraded to >= 1.49.0 to have  https://github.com/grpc/grpc/pull/30567. Once our #39210 is merged I will merge to this PR and CI should work.

File: python/ray/_private/metrics_agent.py
Comment (rkooo567) [inline python/ray/_private/metrics_agent.py:100]: Is there a way to avoid doing this for every metric? IIRC, the regex could be quite expensive when you have lots of data

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Can you merge the latest master? Also lint failure + left of comments

File: src/ray/common/ray_config_def.h
Comment (rynewang) [reply src/ray/common/ray_config_def.h:523]: actually updated and applied to gcs, raylet and core_worker.

Comment (rynewang) [reply src/ray/common/ray_config_def.h:523]: actually a bit awkward, removed raylet and core_worker. Added todo about what's missing

File: src/ray/stats/metric_exporter.cc
Comment (rynewang) [reply src/ray/stats/metric_exporter.cc:141]: added cc test. it's a bit special (has "global" settings) so I put it into a separate cc test file.

File: python/ray/_private/metrics_agent.py
Comment (rynewang) [reply python/ray/_private/metrics_agent.py:103]: done. Now we only do the work if it's grpc.

**Comment by rynewang** [comment]
ready to merge

**Comment by rkooo567** [comment]
Can you update the PR description? Also, test_prometheus_physical_stats_record failure seesm related

**Comment by rynewang** [comment]
updated

**Pull request merged**
